syntax = "proto3";

package darmok.protobuf;

import "math.proto";

message Bgfx {
    enum Attrib {
        Position = 0;
        Normal = 1;
        Tangent = 2;
        Bitangent = 3;
        Color0 = 4;
        Color1 = 5;
        Color2 = 6;
        Color3 = 7;
        Indices = 8;
        Weight = 9;
        TexCoord0 = 10;
        TexCoord1 = 11;
        TexCoord2 = 12;
        TexCoord3 = 13;
        TexCoord4 = 14;
        TexCoord5 = 15;
        TexCoord6 = 16;
        TexCoord7 = 17;
    }

    enum AttribType {
        Uint8 = 0;
        Uint10 = 1;
        Int16 = 2;
        Half = 3;
        Float = 4;
    }
}

message VertexAttribute {
    Bgfx.Attrib bgfx = 1;
    Bgfx.AttribType bgfx_type = 2;
    uint32 num = 3;
    bool normalize = 4;
    bool as_int = 5;
}

message FragmentAttribute {
    Bgfx.Attrib bgfx = 1;
    uint32 num = 2;
    string name = 3;
    Vec4 default_value = 4;
}

message VertexLayout {
    repeated VertexAttribute attributes = 1;
}

message FragmentLayout {
    repeated FragmentAttribute attributes = 1;
}

message Varying {
    VertexLayout vertex = 1;
    FragmentLayout fragment = 2;
    uint32 instance_attribs = 3;
}